File: frontend/src/lib/permissions.ts
export type TenantPermission =
  | "agenda"
  | "clients"
  | "employees"
  | "services"
  | "financial"
  | "cashier"
  | "commissions"
  | "stock"
  | "audit"
  | "settings";

export const tenantRoutePermissions: Record<string, TenantPermission | null> = {
  "/dashboard": null,
  "/dashboard/agenda": "agenda",
  "/dashboard/clients": "clients",
  "/dashboard/employees": "employees",
  "/dashboard/services": "services",
  "/dashboard/financial": "financial",
  "/dashboard/cashier": "cashier",
  "/dashboard/commissions": "commissions",
  "/dashboard/stock": "stock",
  "/dashboard/audit": "audit",
  "/dashboard/settings": "settings",
};

const tenantFallbackOrder: string[] = [
  "/dashboard",
  "/dashboard/agenda",
  "/dashboard/clients",
  "/dashboard/cashier",
  "/dashboard/commissions",
  "/dashboard/employees",
  "/dashboard/services",
  "/dashboard/financial",
  "/dashboard/stock",
  "/dashboard/audit",
  "/dashboard/settings",
];

export function hasTenantPermission(user: any, permission?: TenantPermission | null) {
  if (!permission) return true;
  if (!user) return false;
  if (user.role === "super_admin" || user.role === "admin") return true;
  return !!user.permissions?.[permission];
}

export function getFirstAllowedTenantRoute(user: any) {
  for (const route of tenantFallbackOrder) {
    if (hasTenantPermission(user, tenantRoutePermissions[route])) {
      return route;
    }
  }
  return "/dashboard";
}

File: backend/app/services/permissions.py
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.tenant_settings import TenantSettings
from app.models.user import User

DEFAULT_PERMISSIONS = {
    "gerente": {
        "agenda": True,
        "clients": True,
        "employees": True,
        "services": True,
        "financial": True,
        "cashier": True,
        "commissions": True,
        "stock": True,
        "audit": True,
        "settings": True,
    },
    "recepcionista": {
        "agenda": True,
        "clients": True,
        "employees": False,
        "services": False,
        "financial": False,
        "cashier": True,
        "commissions": False,
        "stock": False,
        "audit": False,
        "settings": False,
    },
    "profissional": {
        "agenda": True,
        "clients": True,
        "employees": False,
        "services": False,
        "financial": False,
        "cashier": False,
        "commissions": True,
        "stock": False,
        "audit": False,
        "settings": False,
    },
}


def get_tenant_permissions(db: Session, tenant_id: str) -> dict[str, dict[str, bool]]:
    settings = db.scalar(select(TenantSettings).where(TenantSettings.tenant_id == tenant_id))
    if settings and settings.permissions:
        return settings.permissions
    return DEFAULT_PERMISSIONS


def get_effective_permissions(db: Session, user: User) -> dict[str, bool]:
    if user.role in {"super_admin", "admin"}:
        modules = {
            key
            for role_permissions in DEFAULT_PERMISSIONS.values()
            for key in role_permissions.keys()
        }
        return {module: True for module in modules}

    if not user.tenant_id:
        return {}

    permissions = get_tenant_permissions(db, user.tenant_id)
    return permissions.get(user.role, {})
